pkg/controller/deployment/deployment_controller.go

```diff
@@ -22,79 +22,314 @@ import (
+	PodRelistPeriod = 5 * time.Minute
 )
 
+// DeploymentController is responsible for synchronizing Deployment objects stored
+// in the system with actual running rcs and pods.
 type DeploymentController struct {
 	client        client.Interface
 	expClient     client.ExtensionsInterface
 	eventRecorder record.EventRecorder
+
+	// To allow injection of syncDeployment for testing.
+	syncHandler func(dKey string) error
+
+	// A store of deployments, populated by the dController
+	dStore cache.StoreToDeploymentLister
+	// Watches changes to all deployments
+	dController *framework.Controller
+	// A store of replication controllers, populated by the rcController
+	rcStore cache.StoreToReplicationControllerLister
+	// Watches changes to all replication controllers
+	rcController *framework.Controller
+	// rcStoreSynced returns true if the RC store has been synced at least once.
+	// Added as a member to the struct to allow injection for testing.
+	rcStoreSynced func() bool
+	// A store of pods, populated by the podController
+	podStore cache.StoreToPodLister
+	// Watches changes to all pods
+	podController *framework.Controller
+	// podStoreSynced returns true if the pod store has been synced at least once.
+	// Added as a member to the struct to allow injection for testing.
+	podStoreSynced func() bool
+
+	// Deployments that need to be synced
+	queue *workqueue.Type
 }
 
-func New(client client.Interface) *DeploymentController {
+// NewDeploymentController creates a new DeploymentController.
+func NewDeploymentController(client client.Interface) *DeploymentController {
 	eventBroadcaster := record.NewBroadcaster()
 	eventBroadcaster.StartLogging(glog.Infof)
 	eventBroadcaster.StartRecordingToSink(client.Events(""))
 
-	return &DeploymentController{
+	dc := &DeploymentController{
 		client:        client,
 		expClient:     client.Extensions(),
 		eventRecorder: eventBroadcaster.NewRecorder(api.EventSource{Component: "deployment-controller"}),
+		queue:         workqueue.New(),
+	}
+
+	dc.dStore.Store, dc.dController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return dc.expClient.Deployments(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return dc.expClient.Deployments(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), options)
+			},
+		},
+		&extensions.Deployment{},
+		FullDeploymentResyncPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc: dc.enqueueDeployment,
+			UpdateFunc: func(old, cur interface{}) {
+				// Resync on deployment object relist.
+				dc.enqueueDeployment(cur)
+			},
+			// This will enter the sync loop and no-op, because the deployment has been deleted from the store.
+			// Note that deleting a controller immediately after scaling it to 0 will not work. The recommended
+			// way of achieving this is by performing a `stop` operation on the deployment.
```

Comment:
No, deployments don't have a stop operation currently. I think @bgrant0607 was advocating over doing this in the server vs in the client where all stop operations are currently implemented.